rust/datafusion/src/physical_plan/expressions.rs

```diff
@@ -2414,6 +2417,212 @@ impl PhysicalSortExpr {
     }
 }
 
+/// InList
+#[derive(Debug)]
+pub struct InListExpr {
+    expr: Arc<dyn PhysicalExpr>,
+    list: Vec<Arc<dyn PhysicalExpr>>,
+}
+
+macro_rules! make_contains {
+    ($ARRAY:expr, $LIST_VALUES:expr, $SCALAR_VALUE:ident, $ARRAY_TYPE:ident) => {{
+        let array = $ARRAY.as_any().downcast_ref::<$ARRAY_TYPE>().unwrap();
+
+        let values = $LIST_VALUES
+            .iter()
+            .map(|expr| match expr {
+                ColumnarValue::Scalar(s) => match s {
+                    ScalarValue::$SCALAR_VALUE(Some(v)) => v,
+                    datatype => unimplemented!("Unexpected type {} for InList", datatype),
+                },
+                ColumnarValue::Array(_) => {
+                    unimplemented!("InList should not receive Array")
+                }
+            })
+            .collect::<Vec<_>>();
+
+        Ok(ColumnarValue::Array(Arc::new(
+            array
+                .iter()
+                .map(|x| x.map(|x| values.contains(&&x)))
```

Comment:
I wonder if this handles `NULL` correctly -- like for a value of where `expr` is NULL the output should be NULL (not true/false). The semantics when there is a literal `NULL` in the inlist are even stranger (but likely could be handled as a follow on PR)

For example:

sqlite> create table t(c1 int);
sqlite> insert into t values (10);
sqlite> insert into t values (20);
sqlite> insert into t values(NULL);
sqlite> select c1, c1 IN (20, NULL) from t;
10|
20|1
|
sqlite> select c1, c1 IN (20) from t;
10|0
20|1
|
```

Note that `10 IN (20, NULL)` is actually `NULL` rather than `FALSE`. Crazy

Reply:
This approach of mapping the array was suggested by @jorgecarleitao when helping me with the StringExpressions: https://github.com/apache/arrow/blob/master/rust/datafusion/src/physical_plan/string_expressions.rs#L84. The benefit is that if the input value is `NULL` (i.e. `None`) then we don't have to do any work on it (the second `map`).

I have confirmed this is the desired behavior against Postgres 13.1 so that any `NULL` input `expr` should return null:

```sql
SELECT NULL IN ('a'); -> NULL
SELECT NULL NOT IN ('a'); -> NULL
SELECT NULL IN (NULL, 'a'); -> NULL
SELECT NULL NOT IN (NULL, 'a'); -> NULL
```

Reply:
As to the second problem, `NULL` in the `list` component it gets even crazier than your example (Postgres 13.1). What a mess.

```sql
SELECT 'a' IN (NULL); -> NULL
SELECT 'a' IN (NULL, 'a'); -> TRUE
SELECT 'a' IN (NULL, 'b'); -> NULL
```